pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -1817,6 +1816,25 @@ impl ResidentLayer {
         }
     }
 
+    /// Read all they keys in this layer which match the ShardIdentity, and write them all to
+    /// the provided writer.  Return the number of keys written.
+    #[tracing::instrument(level = tracing::Level::DEBUG, skip_all, fields(layer=%self))]
+    pub(crate) async fn filter<'a>(
+        &'a self,
+        shard_identity: &ShardIdentity,
+        writer: &mut ImageLayerWriter,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<usize> {
+        use LayerKind::*;
+
+        let key_range = self.layer_desc().key_range.clone();
+
+        match self.downloaded.get(&self.owner.0, ctx).await? {
+            Delta(_) => anyhow::bail!(format!("cannot load_image_keys on a delta layer {self}")),
+            Image(i) => i.filter(key_range, shard_identity, writer, ctx).await,
+        }
+    }
```

Comment:
Additional reason why I added the copy_lsn_prefix into ResidentLayer (should had documented this) is to force the `download_and_keep_resident`. Not forcing the caller to do this will create an avenue for easily doing the download while holding the read lock on layers, which we currently have on (and have always had) on L0 -> L1 compaction.

But `owner: &Arc<LayerInner>` was most likely passed to the previous place of this method, as I noted in my other comment.

Reply:
I'm not sure what you're saying here: this method is already on ResidentLayer, so we are already ensuring that the layer is kept resident

Reply:
Ok, I got confused by the way DownloadedLayer::get is called. Perhaps it has changed?